Add support for joypad touchpads

Nintorch opened the pull request:
- Closes https://github.com/godotengine/godot-proposals/issues/14481

I decided to split my big SDL3 joypad features PR ( https://github.com/godotengine/godot/pull/107967 ) into several smaller PRs, this PR is one of them.

This PR adds the ability to use touchpads on supported controllers, such as DualShock 4, DualSense ~~and (most likely) Steam Deck built-in controls~~ (EDIT: apparently not, since "thirdparty/sdl/joystick/hidapi/SDL_hidapi_steamdeck.c" file doesn't have any references to touchpads).

Currently there's no InputEvent that gets fired when a finger appears/moves on a touchpad.

TODO:
- [x] Mark as experimental
- [x] Add a note about Android automatically turning joypad touchpads into a mouse cursor
- [x] Test on iOS later after https://github.com/godotengine/godot/pull/114316 is merged
- [x] Write a proposal
- [x] Make `p_touchpad` have the default value of `0`
- [x] Allow touchpad events to be ignored if joypad input is disabled on unfocused application
- ~~Input events, including a motion event with movement vector~~ (Can be done in a separate PR if required)
- [x] Method for getting the pressure value
- [x] Default value for `get_joy_touchpad_fingers` method touchpad parameter

TODO for a separate PR:
- [x] Track presses/releases separately from pressure in `joy_touchpad()` method ( https://github.com/godotengine/godot/pull/120578 )
- [x] I think returning `Vector2(-1, -1)` for position and `-1` for pressure would make more sense here. ( https://github.com/godotengine/godot/pull/120578 )
- [x] `InputEventJoypadTouchpadPress` (press/release) ( https://github.com/godotengine/godot/pull/120579 )
- [x] `InputEventJoypadTouchpadMotion` (finger motion, with initial position and velocity to make things easier) ( https://github.com/godotengine/godot/pull/120579 )
- [ ] Tutorial

AThousandShips commented on core/input/input.cpp:
```diff
@@ -731,6 +735,43 @@ Vector3 Input::get_gyroscope() const {
+	}
+
+	TypedArray<int> result;
+	for (auto &i : joy_touch[p_device].touchpad_fingers[p_touchpad]) {
```
`auto` is [not allowed](https://contributing.godotengine.org/en/latest/engine/guidelines/cpp_usage_guidelines.html#auto-keyword)

AThousandShips commented on core/input/input.cpp:
```diff
@@ -731,6 +735,43 @@ Vector3 Input::get_gyroscope() const {
+	if (!joy_touch.has(p_device)) {
+		return 0;
+	}
+	return joy_touch[p_device].num_touchpads;
```
```suggestion
	const TouchpadInfo *touch = joy_touch.getptr(p_device);
	if (!touch) {
		return 0;
	}
	return touch->num_touchpads;
```

AThousandShips commented on core/input/input.h:
```diff
@@ -346,6 +374,10 @@ class Input : public Object {
 
+	void set_joy_features(int p_device, JoypadFeatures *p_features);
+
+	void set_joy_touchpad_finger(int p_device, int p_touchpad, int p_finger, float p_pressure, Vector2 p_value);
```
```suggestion
	void set_joy_touchpad_finger(int p_device, int p_touchpad, int p_finger, float p_pressure, const Vector2 &p_value);
```

AThousandShips commented on core/input/input.h:
```diff
@@ -330,6 +352,12 @@ class Input : public Object {
 
+	Vector2 get_joy_touchpad_finger_position(int p_device, int p_touchpad, int p_finger) const;
+	float get_joy_touchpad_finger_pressure(int p_device, int p_touchpad, int p_finger) const;
+	TypedArray<int> get_joy_touchpad_fingers(int p_device, int p_touchpad) const;
```
```suggestion
	PackedInt32Array get_joy_touchpad_fingers(int p_device, int p_touchpad) const;
```
I'd say

AThousandShips commented on core/input/input.h:
```diff
@@ -174,6 +194,7 @@ class Input : public Object {
 		int mapping = -1;
 		int hat_current = 0;
 		Dictionary info;
+		Input::JoypadFeatures *features;
```
```suggestion
		Input::JoypadFeatures *features = nullptr;
```

AThousandShips commented on core/input/input.cpp:
```diff
@@ -1478,6 +1537,15 @@ void Input::_update_action_cache(const StringName &p_action_name, ActionState &r
+		return;
+	}
+	if (joy_names[p_device].features->get_joy_num_touchpads() > 0) {
+		joy_touch[p_device].num_touchpads = joy_names[p_device].features->get_joy_num_touchpads();
```
```suggestion
	Joypad *joypad = joy_names.getptr(p_device);
	if (!joypad || joypad->features == nullptr) {
		return;
	}
	if (joypad->features->get_joy_num_touchpads() > 0) {
		joy_touch[p_device].num_touchpads = joypad->features->get_joy_num_touchpads();
```

AThousandShips commented on core/input/input.cpp:
```diff
@@ -986,6 +1027,24 @@ void Input::set_joy_axis(int p_device, JoyAxis p_axis, float p_value) {
+	if (!joy_names.has(p_device)) {
+		return;
+	}
+	joy_names[p_device].features = p_features;
```
```suggestion
	Joypad *joypad = joy_names.getptr(p_device);
	if (!joypad) {
		return;
	}
	joypad->features = p_features;
```

AThousandShips commented on doc/classes/Input.xml:
```diff
@@ -147,6 +147,48 @@
+			<description>
+				Returns the number of touchpads on the joypad.
+				[b]Note:[/b] This method may return inaccurate results in case if this joypad is not in Godot's controller database.
+				[b]Note:[/b] This feature is only supported on Windows, Linux and macOS.
```
```suggestion
				[b]Note:[/b] This feature is only supported on Windows, Linux, and macOS.
```
For all

Meorge replied:
```suggestion
				[b]Note:[/b] This feature is only supported on Windows, Linux, and macOS.
```
Oxford comma (maybe nitpicky)

AThousandShips commented on doc/classes/Input.xml:
```diff
@@ -147,6 +147,48 @@
+			<param index="1" name="touchpad" type="int" />
+			<param index="2" name="finger" type="int" />
+			<description>
+				Returns the position of the specified finger on the specified touchpad on the joypad. The X and Y values are in range between 0.0 and 1.0.
```
```suggestion
				Returns the position of the specified finger on the specified touchpad on the joypad. The X and Y values are between [code]0.0[/code] and [code]1.0[/code].
```
Or
```suggestion
				Returns the position of the specified finger on the specified touchpad on the joypad. The X and Y values are the range from [code]0.0[/code] to [code]1.0[/code].
```

Meorge commented on doc/classes/Input.xml:
```diff
@@ -147,6 +147,48 @@
+			<param index="0" name="device" type="int" />
+			<description>
+				Returns the number of touchpads on the joypad.
+				[b]Note:[/b] This method may return inaccurate results in case if this joypad is not in Godot's controller database.
```
```suggestion
				[b]Note:[/b] This method may return inaccurate results if this joypad is not in Godot's controller database.
```

Meorge commented on doc/classes/Input.xml:
```diff
@@ -147,6 +147,48 @@
+			<param index="1" name="touchpad" type="int" />
+			<param index="2" name="finger" type="int" />
+			<description>
+				Returns the pressure of the specified finger on the specified touchpad on the joypad. The pressure values are in range between 0.0 and 1.0.
```
```suggestion
				Returns the pressure of the specified finger on the specified touchpad on the joypad. The pressure values are in range between 0.0 and 1.0.
				If the touchpad doesn't support pressure sensitivity, this method returns [code]1.0[/code].
```
Based on my own testing with DualShock and DualSense controllers

Meorge commented on doc/classes/Input.xml:
```diff
@@ -147,6 +147,48 @@
+			<description>
+				Returns the pressure of the specified finger on the specified touchpad on the joypad. The pressure values are in range between 0.0 and 1.0.
+				If the joypad doesn't have the specified touchpad or the specified finger is not currently touching the touchpad, this method returns [code]0.0[/code].
+				[b]Note:[/b] This feature is only supported on Windows, Linux and macOS.
```
```suggestion
				[b]Note:[/b] This feature is only supported on Windows, Linux, and macOS.
```

Meorge reviewed (approved):
Tested with a wired PS4 DualShock and PS5 DualSense controller on macOS, and it works great!

Here's a very minimal Godot project for testing, adapted from the previous PR, to make it easier for others to test as well: 
[test-joypad-touchpad.zip](https://github.com/user-attachments/files/22982461/test-joypad-touchpad.zip)
(I can put it up on a GitHub repo if necessary, but it's so minimal unlike the adaptive triggers one that it's probably fine this way 😅 )

AThousandShips commented on doc/classes/Input.xml:
```diff
@@ -147,6 +147,49 @@
+			<param index="1" name="touchpad" type="int" />
+			<param index="2" name="finger" type="int" />
+			<description>
+				Returns the position of the specified finger on the specified touchpad on the joypad. The X and Y values are in the range 0.0 to 1.0, with [code](0.0, 0.0)[/code] representing the top left corner of the touchpad and [code](1.0, 1.0)[/code] representing the bottom right corner.
```
```suggestion
				Returns the position of the specified finger on the specified touchpad on the joypad. The X and Y values are in the range from [code]0.0[/code] to [code]1.0[/code], with [code](0.0, 0.0)[/code] representing the top left corner of the touchpad and [code](1.0, 1.0)[/code] representing the bottom right corner.
```

AThousandShips commented on doc/classes/Input.xml:
```diff
@@ -147,6 +147,49 @@
+			<param index="1" name="touchpad" type="int" />
+			<param index="2" name="finger" type="int" />
+			<description>
+				Returns the pressure of the specified finger on the specified touchpad on the joypad. The pressure values are in range between 0.0 and 1.0.
```
```suggestion
				Returns the pressure of the specified finger on the specified touchpad on the joypad. The pressure values are in the range between [code]0.0[/code] and [code]1.0[/code].
```

Nintorch commented on core/input/input.h:
```diff
@@ -158,6 +156,19 @@ class Input : public Object {
+	struct TouchpadInfo {
+		int num_touchpads = 0;
+		// The first int index refers to touchpad ID, the second one refers to finger ID on that touchpad.
+		HashMap<int, HashMap<int, TouchpadFingerInfo>> touchpad_fingers;
```
It probably makes sense for it to be
```suggestion
		HashMap<Vector2i, TouchpadFingerInfo> finger_info;
```
Alongside with `Vector<int> fingers`

Nintorch commented on drivers/sdl/joypad_sdl.h:
```diff
@@ -62,6 +62,8 @@ class JoypadSDL {
 		bool supports_force_feedback = false;
 		uint64_t ff_effect_timestamp = 0;
 
+		int get_joy_num_touchpads() const override;
```
```suggestion
		virtual int get_joy_num_touchpads() const override;
```

AThousandShips commented on core/input/input.cpp:
```diff
@@ -1016,6 +1076,18 @@ bool Input::has_joy_light(int p_device) const {
+void Input::set_joy_touchpad_finger(int p_device, int p_touchpad, int p_finger, const Vector2 &p_value, float p_pressure) {
+	_THREAD_SAFE_METHOD_
+	TouchpadInfo &touch = joy_touch[p_device];
+	Vector2i index = { p_touchpad, p_finger };
```
```suggestion
	Vector2i index(p_touchpad, p_finger);
```
Standard syntax

AThousandShips commented on core/input/input.h:
```diff
@@ -149,6 +156,19 @@ class Input : public Object {
 	HashMap<int, VibrationInfo> joy_vibration;
 
+	struct TouchpadFingerInfo {
+		Vector2 position = Vector2();
```
```suggestion
		Vector2 position;
```

Nintorch committed:
Add support for joypad touchpads

This PR adds the ability to use touchpads on supported controllers, such as DualShock 4 and DualSense.
Currently there's no InputEvent that gets fired when a finger appears/moves on a touchpad.

bruvzg reviewed (approved):
Tested on macOS and iOS with DualSense controller, controller touchpad seems to be working as expected.

Nintorch commented:
Thank you for your review, bruvzg! :)

Nintorch commented on core/input/input.h:
```diff
@@ -385,6 +395,11 @@ class Input : public Object {
 	Vector3 get_gyroscope() const;
 
+	Vector2 get_joy_touchpad_finger_position(int p_device, int p_touchpad, int p_finger) const;
+	PackedInt32Array get_joy_touchpad_fingers(int p_device, int p_touchpad) const;
```
Should these methods be modified to make the `p_touchpad` parameter default to `0` since (I assume) these methods will most likely be used on DualShock 4/DualSense controllers that have only one touchpad?

```suggestion
	Vector2 get_joy_touchpad_finger_position(int p_device, int p_finger, int p_touchpad = 0) const;
	PackedInt32Array get_joy_touchpad_fingers(int p_device, int p_touchpad = 0) const;
```

Meorge replied:
I think that makes sense! More generally, if someone doesn't bother to specify which touchpad they want to use, I think it's reasonable to assume that they want *the* touchpad they must believe exists for the device.

Birdulon commented:
> Apparently, SDL only reports pressure values of 0 or 1, so I will probably need to remove the method to get the pressure value, because it's not that useful in this case. <img alt="image" width="1457" height="457" src="https://private-user-images.githubusercontent.com/92302738/538748353-5f2b459f-9f0e-41cf-9f7f-49cfcb52e431.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzg2NzA3NTMsIm5iZiI6MTc3ODY3MDQ1MywicGF0aCI6Ii85MjMwMjczOC81Mzg3NDgzNTMtNWYyYjQ1OWYtOWYwZS00MWNmLTlmN2YtNDljZmNiNTJlNDMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNTEzVDExMDczM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllZTZjMjY0ZDkyNGJiMDJhY2E0MzQ1OTA5OGUwYTg5YmRkYmY2MDU5ZWYzMzg5OGQ3MDQyZjE5M2Q4Y2EwMmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.6sa77Vbb15IWiPJpisRWzkqdtbznuotBLpk5v5D_3as">

True for Sony controllers, however Valve controllers do report a range of pressure values. Tested on Steam Deck Desktop Mode (Gaming Mode forces Steam Input to present an emulated xinput controller) on SDL3 `main`, and the new Steam Controller (codename "triton") with https://github.com/libsdl-org/SDL/pull/15528
<img width="2044" height="1416" alt="image" src="https://github.com/user-attachments/assets/49a28d11-063e-44c7-96a7-c67b188e24d9" />

AThousandShips reviewed (approved):
Style and docs look good, haven't tested

Nintorch commented:
Thank you for your review, AThousandShips! :)

Birdulon commented:
Messing around again now that I have a project to properly use the Steam Controller in. I've put up a branch roughly tracking this one with very hackily backported Steam Controller (2026 "Triton") support from SDL 3.4.10 for anyone else with a Steam Controller interested in playing around in the meantime.

First observation for this PR: Zero pressure `SDL_EVENT_GAMEPAD_TOUCHPAD_DOWN` and `SDL_EVENT_GAMEPAD_TOUCHPAD_MOTION` events are legitimate and very desirable. I've patched the logic to not use pressure as a heuristic and instead use `SDL_EVENT_GAMEPAD_TOUCHPAD_UP` to actually release the finger: https://github.com/Birdulon/godot/commit/81c76e9576c839cab925ee069a1f06d4d0cde275
I can't physically obtain a (0,0) touch on my controller but I do have reservations about `Vector2.ZERO` being the guard value for `get_joy_touchpad_finger_position` and `0.0` for `get_joy_touchpad_finger_pressure`. Of course, `get_joy_touchpad_fingers` can be used to confirm that a finger is not being tracked, but there is the technical possibility of (x=0, y=0, pressure=0) being a valid finger state mistakenly treated as absent, and that bothers me.

Second observation: `SDL_EVENT_GAMEPAD_TOUCHPAD_UP` could rightly be considered a useful event in and of itself, like releasing a button. I haven't yet tested if it properly includes the coordinates of the release, because...
Maybe instead of only having functions to poll the current state, it could also fire off Events like the computer touchpad does? I know this increases the scope of the PR and could be handled in a future one, but now seems like a good time to talk API decisions.

Thanks @Nintorch for your work on this PR and also on #106218, I really appreciate being able to play around with this stuff and I hope my comments here and elsewhere don't come off as complaining too much!

Nintorch commented:
Hello! Thank you for your feedback!

> First observation for this PR: Zero pressure `SDL_EVENT_GAMEPAD_TOUCHPAD_DOWN` and `SDL_EVENT_GAMEPAD_TOUCHPAD_MOTION` events are legitimate and very desirable. I've patched the logic to not use pressure as a heuristic and instead use `SDL_EVENT_GAMEPAD_TOUCHPAD_UP` to actually release the finger: [Birdulon@81c76e9](https://github.com/Birdulon/godot/commit/81c76e9576c839cab925ee069a1f06d4d0cde275)

I didn't know it was possible, but in this case I agree that it is a good idea to track presses/releases separately from pressure. I think we can add these features in a separate PR, since this one is already marked as ready to merge for 4.8. 😅 

> I can't physically obtain a (0,0) touch on my controller but I do have reservations about `Vector2.ZERO` being the guard value for `get_joy_touchpad_finger_position` and `0.0` for `get_joy_touchpad_finger_pressure`. Of course, `get_joy_touchpad_fingers` can be used to confirm that a finger is not being tracked, but there is the technical possibility of (x=0, y=0, pressure=0) being a valid finger state mistakenly treated as absent, and that bothers me.

Good idea, I think returning `Vector2(-1, -1)` for position and `-1` for pressure would make more sense here.

> Second observation: `SDL_EVENT_GAMEPAD_TOUCHPAD_UP` could rightly be considered a useful event in and of itself, like releasing a button. I haven't yet tested if it properly includes the coordinates of the release, because... Maybe instead of only having functions to poll the current state, it could also fire off Events like the computer touchpad does? I know this increases the scope of the PR and could be handled in a future one, but now seems like a good time to talk API decisions.

This is planned to be done in a separate PR! :)

> Thanks @Nintorch for your work on this PR and also on #106218, I really appreciate being able to play around with this stuff and I hope my comments here and elsewhere don't come off as complaining too much!

You're welcome! And I'm glad my PRs are useful, that means a lot! :) ❤️

Birdulon commented on core/input/input.cpp:
```diff
@@ -1823,6 +1888,22 @@ void Input::joy_motion_sensors(int p_device, const Vector3 &p_accelerometer, con
 	motion->gamepad_motion->ProcessMotion(gyro_degrees.x, gyro_degrees.y, gyro_degrees.z, accel_g.x, accel_g.y, accel_g.z, delta_time);
 }
 
+void Input::joy_touchpad(int p_device, int p_touchpad, int p_finger, const Vector2 &p_value, float p_pressure) {
```
`p_value` here is `p_position` in the header

Nintorch replied:
Thanks for catching that, I should be able to fix it today!

Birdulon commented:
> This PR adds the ability to use touchpads on supported controllers, such as DualShock 4, DualSense ~and (most likely) Steam Deck built-in controls~ (EDIT: apparently not, since "thirdparty/sdl/joystick/hidapi/SDL_hidapi_steamdeck.c" file doesn't have any references to touchpads).

Forgot to clarify on this earlier - SDL did get Steam Deck touchpad support after that was written. It requires Steam to not be running, hence my earlier "only in Desktop Mode" comment. Desktop Mode with Steam running, and Gaming Mode both hijack the controller (Deck built-in *and* external Steam Controllers!) and make it impossible to access the non-xinput/kbm controls through SDL.
The only option for Steam games is to use SteamInput API which
1) AFAIK requires license stuff prohibiting it from ever being upstreamed in Godot unless Valve decides to change it, and
2) doesn't let you play with the touchpads or extra buttons directly, just allows you to define "actions" that Steam will allow mapping the fancy inputs to *outside* of your game. Your game has the privilege of setting the default mappings, and asking what is mapped to what, so that you can display the correct glyphs, but you don't get to ask what the state of the touchpads or gyroscope or accelerometers are directly, as far as I can tell.

So anyone shipping a game on multiple platforms will need to handle the input at least slightly differently for their Steam release (https://godotsteam.com/tutorials/inputs/).



Random use case feedback: I'm finding myself operating on Vector3(x, y, pressure). Convenient for deltas and some 3D input reasoning, but perhaps too niche to consider adding it to the API.

Calinou commented on doc/classes/Input.xml:
```diff
@@ -217,6 +217,47 @@
+			<param index="1" name="finger" type="int" />
+			<param index="2" name="touchpad" type="int" default="0" />
+			<description>
+				Returns the position of the specified finger on the specified touchpad on the joypad. The X and Y values are in the range from [code]0.0[/code] to [code]1.0[/code], with [code](0.0, 0.0)[/code] representing the top left corner of the touchpad and [code](1.0, 1.0)[/code] representing the bottom right corner.
```
```suggestion
				Returns the position of the specified finger on the specified touchpad on the joypad. The X and Y values are in the range from [code]0.0[/code] to [code]1.0[/code], with [code](0.0, 0.0)[/code] representing the top-left corner of the touchpad and [code](1.0, 1.0)[/code] representing the bottom-right corner.
```

Calinou commented on doc/classes/Input.xml:
```diff
@@ -217,6 +217,47 @@
+			<param index="1" name="touchpad" type="int" default="0" />
+			<description>
+				Returns an array of finger IDs that are currently touching the specified touchpad on the joypad.
+				If the joypad doesn't have the specified touchpad or no fingers are currently touching this touchpad, this method returns an empty array.
```
Merge to a single paragraph to save space:

```suggestion
				Returns an array of finger IDs that are currently touching the specified touchpad on the joypad. If the joypad doesn't have the specified touchpad or no fingers are currently touching this touchpad, this method returns an empty array.
```

Calinou reviewed (approved):
Tested locally on DualSense and DualSense Edge on Linux (both wired), it works as expected. Code and documentation look good to me (aside of small stylistic changes below).

**Testing project:** [test_pr_111714.zip](https://github.com/user-attachments/files/28645677/test_pr_111714.zip)

Nintorch commented:
Thank you for your review, Calinou! :)

Repiteo commented:
Thanks!
